src/libsync/discovery.cpp

```diff
@@ -0,0 +1,1218 @@
+                        item->_instruction = up._modtime == localEntry.modtime && up._size == localEntry.size
+                            ? CSYNC_INSTRUCTION_UPDATE_METADATA : CSYNC_INSTRUCTION_SYNC;
+                        item->_direction = SyncFileItem::Up;
+
+                        // Update the etag and other server metadata in the journal already
+                        // (We can't use a typical CSYNC_INSTRUCTION_UPDATE_METADATA because
+                        // we must not store the size/modtime from the file system)
+                        OCC::SyncJournalFileRecord rec;
+                        if (_discoveryData->_statedb->getFileRecord(path._original, &rec)) {
+                            rec._path = path._original.toUtf8();
+                            rec._etag = serverEntry.etag;
+                            rec._fileId = serverEntry.fileId;
+                            rec._modtime = serverEntry.modtime;
+                            rec._type = item->_type;
+                            rec._fileSize = serverEntry.size;
+                            rec._remotePerm = serverEntry.remotePerm;
+                            rec._checksumHeader = serverEntry.checksumHeader;
+                            _discoveryData->_statedb->setFileRecordMetadata(rec);
+                        }
+                    } else {
+                        item->_instruction = CSYNC_INSTRUCTION_CONFLICT;
+                        item->_direction = SyncFileItem::None;
+                    }
+                } else {
+                    // If the size or mtime is different, it's definitely a conflict.
+                    bool isConflict = (serverEntry.size != localEntry.size) || (serverEntry.modtime != localEntry.modtime);
+
+                    // SO: If there is no checksum, we can have !is_conflict here
+                    // even though the files have different content! This is an
+                    // intentional tradeoff. Downloading and comparing files would
+                    // be technically correct in this situation but leads to too
+                    // much waste.
+                    // In particular this kind of NEW/NEW situation with identical
+                    // sizes and mtimes pops up when the local database is lost for
+                    // whatever reason.
+                    item->_instruction = isConflict ? CSYNC_INSTRUCTION_CONFLICT : CSYNC_INSTRUCTION_UPDATE_METADATA;
+                    item->_direction = isConflict ? SyncFileItem::None : SyncFileItem::Down;
+                }
+            }
+            if (dbEntry._type == ItemTypeVirtualFile)
+                item->_type = ItemTypeVirtualFileDownload;
+            if (item->_file.endsWith(_discoveryData->_syncOptions._virtualFileSuffix)) {
+                item->_file.chop(_discoveryData->_syncOptions._virtualFileSuffix.size());
+                item->_type = ItemTypeVirtualFileDownload;
+            }
```

Comment:
These two ifs are intended to catch cases where the user replaces a virtual file with a real one and we produce a conflict? I'm confused.

Reply:
exactly.
You create a file with the same name as an existing file, this will cause the file to be made as a conflict and the actual file to be downloaded.
This is the same behavior as before and is tested by `testVirtualFileConflict`